clang/tools/libclang/CDependencies.cpp

```diff
@@ -146,38 +157,45 @@ static CXFileDependencies *getFullDependencies(
     MDC(Context, MDS);
   }
 
-  const FullDependencies &FD = FDR.FullDeps;
-  CXFileDependencies *FDeps = new CXFileDependencies;
-  FDeps->ContextHash = cxstring::createDup(FD.ID.ContextHash);
-  FDeps->FileDeps = cxstring::createSet(FD.FileDeps);
-  std::vector<std::string> Modules;
-  for (const ModuleID &MID : FD.ClangModuleDeps)
-    Modules.push_back(MID.ModuleName + ":" + MID.ContextHash);
-  FDeps->ModuleDeps = cxstring::createSet(Modules);
-  FDeps->BuildArguments = cxstring::createSet(FD.DriverCommandLine);
-  return FDeps;
+  HandleFullDeps(std::move(FDR.FullDeps));
+  return true;
 }
 
-static CXFileDependencies *
+static bool
 getFileDependencies(CXDependencyScannerWorker W, int argc,
                     const char *const *argv, const char *WorkingDirectory,
                     CXModuleDiscoveredCallback *MDC, void *Context,
                     CXString *error, LookupModuleOutputCallback LookupOutput,
-                    llvm::Optional<StringRef> ModuleName = None) {
-  if (!W || argc < 2)
-    return nullptr;
+                    bool DeprecatedDriverCommand,
+                    llvm::Optional<StringRef> ModuleName,
+                    HandleFullDepsCallback HandleFullDeps) {
+  if (!W) {
+    if (error)
+      *error = cxstring::createDup("missing CXDependencyScannerWorker");
```

Comment:
I like the fact that we're providing more context here. Do you think it would be reasonable to upgrade these messages into proper diagnostics when we update the API to return `CXDiagnosticSet` instead of `CXString`?

Reply:
I'm not sure how reasonable it is to treat API errors as diagnostics -- are there other examples that behave that way in libclang? We could instead follow `clang_createTranslationUnit` and have the main return value be `CXErrorCode::CXError_InvalidArguments` with no diagnostics.  That's not as specific as this string, but it's better than just success/fail at least.

Reply:
I didn't see examples of treating API errors as diagnostics, no. The approach taken in `clang_createTranslationUnit` seems like a good compromise though. I think it would be better to use that approach in this PR instead of reusing `error`. That will make switching to `CXDiagnosticSet` later easier.

Reply:
Done in latest changes.